### PR TITLE
Backport fixes for UPD-64 hotfix

### PR DESCRIPTION
**Pull request by mseri**
This includes:

* CA-252876
* CA-250757
* CA-251251
* CA-247399
* CA-247452
* CA-245389

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

### CODE REVIEW

**Comment by mseri** [comment]
The compilation is successful and passes the `make test` step. All the commits are simple cherry picks, the only one that needed backport was https://github.com/xapi-project/xen-api/pull/3053/commits/5e717ca0a84d3aac1c6bf6b9616dad4bb3716f90 and I backported also the test to make sure I wasn't messing up

File: ocaml/auth/extauth_plugin_ADpbis.ml [hunk 1]
```diff
@@ -30,21 +42,8 @@ struct
 
   let user_friendly_error_msg = "The Active Directory Plug-in could not complete the command. Additional information in the logs."
 
-  open Stdext.Xstringext
```
Comment (Frezzle) [inline ocaml/auth/extauth_plugin_ADpbis.ml]: Well spotted; maybe we should take this out from master too?

Comment (mseri) [inline ocaml/auth/extauth_plugin_ADpbis.ml]: It's moved above. Not removed.

**Comment by Frezzle** [comment]
All backports look faithful to their originals.

File: ocaml/auth/extauth_plugin_ADpbis.ml (hunk 1 above)
Comment (Frezzle) [inline ocaml/auth/extauth_plugin_ADpbis.ml]: Yes, but I mean it is duplicated in master too right now (https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/extauth_plugin_ADpbis.ml), just as you have corrected here.

Comment (mseri) [inline ocaml/auth/extauth_plugin_ADpbis.ml]: :O I missed that when I reviewed the PR. Yep, I will make a PR for master later

**Pull request merged**